src/slurm_plugin/clustermgtd.py

```diff
@@ -0,0 +1,624 @@
+
+    def _is_node_healthy(self, node, private_ip_to_instance_map):
+        """Check if a slurm node is considered healthy."""
+        if node.is_static_node():
+            return (
+                ClusterManager._is_static_node_configuration_valid(node)
+                and ClusterManager._is_backing_instance_valid(
+                    node, instance_ips_in_cluster=list(private_ip_to_instance_map.keys())
+                )
+                and self._is_node_state_healthy(node, private_ip_to_instance_map)
+            )
+        else:
+            return ClusterManager._is_backing_instance_valid(
+                node, instance_ips_in_cluster=list(private_ip_to_instance_map.keys())
+            ) and self._is_node_state_healthy(node, private_ip_to_instance_map)
+
+    @staticmethod
+    def _handle_unhealthy_dynamic_nodes(unhealthy_dynamic_nodes):
+        """
+        Maintain any unhealthy dynamic node.
+
+        Setting node to down will let slurm requeue jobs allocated to node.
+        Setting node to power_down will terminate backing instance and reset dynamic node for future use.
+        """
+        log.info("Setting unhealthy dynamic nodes to down and power_down: %s", unhealthy_dynamic_nodes)
+        set_nodes_down_and_power_save(
+            [node.name for node in unhealthy_dynamic_nodes], reason="Schduler health check failed"
+        )
+
+    def _handle_unhealthy_static_nodes(self, unhealthy_static_nodes, private_ip_to_instance_map):
+        """
+        Maintain any unhealthy static node.
+
+        Set node to down, terminate backing instance, and launch new instance for static node.
+        """
+        log.info(
+            "Setting following unhealthy static nodes to DOWN and terminating associated instances: %s",
+            unhealthy_static_nodes,
+        )
+        node_list = [node.name for node in unhealthy_static_nodes]
+        try:
+            # Set nodes into down state so jobs can be requeued immediately
+            set_nodes_down(node_list, reason="Static node maintenance: unhealthy node is being replaced")
+        except Exception as e:
+            logging.error(
+                "Error setting unhealthy static nodes into down state, continuing with instance replacement: %s", e,
+            )
+        instances_to_terminate = []
+        for node in unhealthy_static_nodes:
+            backing_instance = private_ip_to_instance_map.get(node.nodeaddr)
+            if backing_instance:
+                instances_to_terminate.append(backing_instance.id)
+        log.info(
+            "Terminating instances %s that are backing unhealthy static nodes: %s",
+            instances_to_terminate,
+            unhealthy_static_nodes,
+        )
+        self.instance_manager.delete_instances(
+            instances_to_terminate, terminate_batch_size=self.sync_config.terminate_max_batch_size
+        )
+        log.info("Launching new instances for unhealthy static nodes: %s", unhealthy_static_nodes)
```

Comment:
I'm leaving it here but as a general comment we should review the logs cause I'm afraid they are really verbose and significantly grow with the cluster size